pkg/azuredx/azureauth/auth.go

```diff
@@ -0,0 +1,133 @@
+		return "", nil
+	}
+
+	// Azure requires an ID token (instead of an access token) for the exchange.
+	userToken, ok := req.Headers["X-ID-Token"]
+	if !ok {
+		if _, ok := req.Headers["Authorization"]; !ok {
+			// Grafana permits a mixture of OAuth2 and system accounts.
+			backend.Logger.Warn("user credentials unavailable for data request; continue with service principal despite on-behalf-of configuration", "user", user.Login)
+			return "", nil
+		}
+
+		backend.Logger.Error("ID token absent for data request; enable them on the Azure portal under: “App Registrations” → the respective application → “Manage” → “Authentication”", "user", user.Login)
+		return "", errors.New("ID token absent for data request")
+	}
+
+	onBehalfOfToken, err := c.onBehalfOf(userToken)
+	if err != nil {
+		backend.Logger.Error(err.Error(), "user", user.Login)
+		// Don't leak any context to the end-user.
+		return "", errors.New("on-behalf-of token exchange failed for data request")
+	}
+	backend.Logger.Debug("aquired on-behalf-of token for data request")
+
+	return "Bearer " + onBehalfOfToken, nil
+}
+
+// OnBehalfOf resolves a token which impersonates the subject of userToken.
+// UserToken has to be an ID token. See the following link for more detail.
+// https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow
+func (c *ServiceCredentials) onBehalfOf(userToken string) (onBehalfOfToken string, err error) {
+	params := make(url.Values)
+	params.Set("client_id", c.ClientID)
+	params.Set("client_secret", c.Secret)
+	params.Set("grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer")
+	params.Set("assertion", userToken)
+	params.Set("scope", c.ClusterURL+"/.default")
+	params.Set("requested_token_use", "on_behalf_of")
+
+	// https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-protocols#endpoints
+	tokenURL := "https://login.microsoftonline.com/" + url.PathEscape(c.TenantID) + "/oauth2/v2.0/token"
+	req, err := http.NewRequest("GET", tokenURL, strings.NewReader(params.Encode()))
+	if err != nil {
+		return "", fmt.Errorf("on-behalf-of grant request <%q> instantiation: %w", tokenURL, err)
+	}
+	req.Header.Set("Accept", "application/json")
+
+	// HTTP Exchange
+	reqStart := time.Now()
+	resp, err := c.PostForm(tokenURL, params)
+	if err != nil {
+		return "", fmt.Errorf("on-behalf-of grant POST <%q>: %w", tokenURL, err)
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("on-behalf-of grant POST <%q> response: %w", tokenURL, err)
+	}
+	onBehalfOfLatencySeconds.WithLabelValues(strconv.Itoa(resp.StatusCode)).Observe(float64(time.Since(reqStart)) / float64(time.Second))
+	if resp.StatusCode/100 != 2 {
+		var deny struct {
+			Desc string `json:"error_description"`
+		}
+		_ = json.Unmarshal(body, &deny)
+		return "", fmt.Errorf("on-behalf-of grant POST <%q> status %q: %q", tokenURL, resp.Status, deny.Desc)
+	}
```

Comment:
I would remove this metric from this PR

Reply:
Metrics are an integral part of the functionality just like logging is. We can take it out of course if you really prefer an extra step.

Reply:
I know but this is not working atm and if we want to add metrics it would be nice to follow some more metrics, not just this one. So yes please take it out for the moment, we can create another issue to include metrics.

Reply:
That makes sense. No untested stuff in the commit.